edgedns/client.go

```diff
@@ -0,0 +1,110 @@
+package edgedns
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/corbaltcode/go-akamai"
+	"github.com/corbaltcode/go-akamai/internal/request"
+)
+
+const defaultTTL = 300
+
+const (
+	RecordTypeCNAME = "CNAME"
+	RecordTypeTXT   = "TXT"
+)
+
+type TypesResponse struct {
+	Types []string `json:"types"`
+}
+
+type Recordset struct {
+	Name  string   `json:"name"`
+	Type  string   `json:"type"`
+	TTL   int      `json:"ttl"`
+	Rdata []string `json:"rdata"`
+}
+
+type RecordsetQueryArgs struct {
+	Page     int
+	PageSize int
+	Search   string
+	ShowAll  bool
+	SortBy   string
+	Types    string
+}
+
+type RecordsetResponse struct {
+	Metadata   MetadataH   `json:"metadata"`
+	Recordsets []Recordset `json:"recordsets"`
+}
+
+type MetadataH struct {
+	LastPage      int  `json:"lastPage"`
+	Page          int  `json:"page"`
+	PageSize      int  `json:"pageSize"`
+	ShowAll       bool `json:"showAll"`
+	TotalElements int  `json:"totalElements"`
+}
+
+type Client struct {
+	Credentials akamai.Credentials
+}
+
+func handleQueryArgs(q url.Values) string {
+	s := q.Encode()
+	if len(s) == 0 {
+		return s
+	}
+	return "?" + s
+}
+
+func (c *Client) ListRecordsets(zone string, queryArgs url.Values) (*RecordsetResponse, error) {
```

Comment:
The goal of the client library is to make the API feel more like the language, so IMO, it's better not to expose HTTP-specific details like `url.Values`. Also, details of Akamai's protocol could be made to feel more Go-like, and unimportant details could be hidden.

One possible interface, where all but the first argument is optional (can be the zero value):

```
func (c *Client) GetRecordsets(zone string, recordsetTypes []RecordsetType, sortFields []string, search string) ([]Recordset, error)
```

If pagination is necessary, maybe it'd be worth implementing a paginator so you could do:

```
p := edgedns.NewGetRecordsetsPaginator(zone, ...)

for p.HasMorePages() {
  page, err := p.NextPage()
  ...
}
```

The library would deal with the details of paginating so the client can just do the standard Go thing. (Maybe some common pagination stuff could be added to the `requests` package.)

Reply:
I think I agree with you fundamentally on this if we were trying to implement a general use-case API, but IMO, as far as ADP itself is concerned, since it should be proxying the akamai API and providing relatively minimal interceptions otherwise, I was trying to map a request on an input directly to the same request on the output. If the request came in with args, those args need to map to the backend. I think that's why I started by exposing the `url.Values`.

I like the idea of abstracting those details away, and as long as we can provide functional pass-through of the needed operations, then I'm fully on board with making it Go-like